### DISCUSSION TITLE
Do I need to return something from a page method in order to ensure correct enqueuing?

### DISCUSSION THREAD

**Discussion by ZachMcCorkhill** [cypress-io/cypress#18049] [Questions and Help]
Setting aside hopefully the discussion of whether to use Page Object Models, which we use to a certain extent, I've seen some intermittent failures where it looks like steps are skipped in a test. I can't reproduce it with an example (and it's very intermittent), but what I noticed is that one of the methods our page class returns void, and those are the steps that appear to be skipped. Our code is something like this:

class Page {
    static get box() { return cy.get('#box'); }

    static doA() {
        return this.box.click();
    }

    static doB() {
        this.doC();
    }

    static doC() {
        return cy.get('#filter').type('search');
    }
}

...

it('test', () => {
        Page.doB();
        Page.doA();
});

...

There's a lot more going on than that, but hopefully that illustrates the bones of it. Most of the time this works fine, doB() runs followed by doA(). But occasionally the test fails and in the test runner screenshot I see that doA() is called and fails because doB() was never called. I don't return what doB does because I don't need/want to chain off of it. Do I need to return it anyway to ensure the commands enqueue in the correct order?

**Comment by bahmutov** [comment]
I have seen it too in the interactive mode and I don’t think it depends on the page object. Once in a while a longer test “completes” but without running all commands - just 1/2 or 3/4 It is extremely intermittent. I would add a check to the cypress logic to verify the number of completed commands vs queued commands. Something must be wrong and the test believes it is done way before all queued commands are processed.

**Reply by ZachMcCorkhill** [reply]
I've noticed that same thing in interactive mode where not all commands are run and the test finishes, but it's always still in the correct order, so I didn't think it was related to what I'm seeing with this issue. For this issue, it's a command that is skipped and then a subsequent command is run. Do you still think they could be related?

**Comment by bahmutov** [comment]
Hmm these might be related - if there is an error in the logic of commands queue then it might possibly skip a command. But are you sure the first command is skipped and not just flaky because the action B could not be executed because the app is not ready? Like does it show the command B in the command log?

**Reply by ZachMcCorkhill** [reply]
It does not show command B in the command log. Action B does some cy.get() style stuff, though, so I would think if the app wasn't ready it would fail.

**Comment by romankhomitskyi** [comment]
I have more complicated Page Object and without issues so far

**Reply by romankhomitskyi** [reply]
In general, I have similar methods but they are wrapped a bit with extra logic

**Reply by romankhomitskyi** [reply]
![IMG_20210913_212559_273](https://user-images.githubusercontent.com/57874507/133136890-89483974-73b2-41ea-9e6f-b6430ad3720d.jpg)

**Reply by romankhomitskyi** [reply]
![IMG_20210913_212549_279](https://user-images.githubusercontent.com/57874507/133137221-3ea0797a-c5a9-467f-b8ab-5424f102edd4.jpg)

**Reply by ZachMcCorkhill** [reply]
Yes, this looks similar to my setup. So my question is in your resetValue() method, for example, if you didn't have `**return** this.control.click()....` and instead just had `this.control.click()....`? Is it possible with the way Cypress enqueues scripts that if your test then was written
`Dropdown.CyControl.resetValue();`
`Dropdown.CyControl.setValue('foo');`
that .setValue() could be called ahead of resetValue()? I.e. does Cypress enqueue methods that have a return value differently than functions that don't, even if both functions are calling Cypress chainables (such as cy.get())?

**Reply by romankhomitskyi** [reply]
I don' think so
I have a method that selects action from dropdown
(Edit, Delete) it returns void
After this method I call methods from EditPage, or DeletePage which either return this (current page) or new page ( Page A), also could return void
Everything is ok
